File: Problem Set Volumes/Volume 1 (100-199)/UVa_113_Power_of_Cryptography.cpp

/*
	UVa 113 - Power of Cryptography

	To build using Visucal Studio 2008:
		cl -EHsc -O2 UVa_113_Power_of_Cryptography.cpp

	An accepted solution.
*/

/*
	For k^n = p, getting the logarithm of both sides results in:
		n * log(k) = log(p).
		k = exp(log(p) / n).

	Or,

	For k^n = p,
		k = p^(1 / n) = pow(p, 1 / n)
*/

#include <iostream>
#include <cstdio>
#include <cmath>
using namespace std;

int main()
{
	double n;
	while (cin >> n) {
		double p;
		cin >> p;
		printf("%.lf\n", pow(p, 1 / n));
	}
	return 0;
}

File: Problem Set Volumes/Volume 1 (100-199)/UVa_113_Power_of_Cryptography.java

// Uva 113 - Power of Cryptography
// Another accepted solution.

/*
	For k^n = p, getting the logarithm of both sides results in:
		n * log(k) = log(p).
		k = exp(log(p) / n).
*/

import java.util.Scanner;
import java.math.BigInteger;

class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		while (sc.hasNext()) {
			int n = sc.nextInt();
			BigInteger p = sc.nextBigInteger();
			int k = (int)(Math.exp(Math.log(p.doubleValue()) / n));
			while (p.compareTo(BigInteger.valueOf(k).pow(n)) > 0)
				k++;
			System.out.println(k);
		}
	}
}

